# Load libraries ----------------------------------------------------------

library(tidyverse)
library(readxl)
library(here)
library(GenomicFeatures)
library(GenomicRanges)
library(gridExtra)

# Arguments ---------------------------------------------------------------

options(warn=-1)

args <-
  list(
    iPSC = list(
      path_to_gff = here::here("data", "iPSC", "iPSC_charlie_corrected.gtf.cds.gff"),
      path_to_sqanti = here::here("data", "iPSC", "iPSC_charlie_classification.txt"),
      path_to_samples = here::here("data", "iPSC", "Charlies_samples.xlsx"),
      path_to_reads = here::here("data", "iPSC", "reads_per_sample.txt")),
    Brain = list(
      path_to_gff = here::here("data", "Brain", "Brain_regions_corrected.gtf.cds.gff"),
      path_to_sqanti = here::here("data", "Brain", "Brain_regions_classification.txt"))
  )

# Load data ---------------------------------------------------------------

# GFF file from SQANTI3
Isoforms_iPSC <- 
  rtracklayer::import(args$iPSC$path_to_gff)

Isoforms_Brain <- 
  rtracklayer::import(args$Brain$path_to_gff)

# Transcript classifications generated by SQANTI3
SQANTI_classification_iPSC <- 
  read_tsv(args$iPSC$path_to_sqanti, 
           show_col_types = F)

SQANTI_classification_Brain <- 
  read_tsv(args$Brain$path_to_sqanti, 
           show_col_types = F)

# Sample information from Charlie Arber
Samples <-
  read_excel(args$iPSC$path_to_samples) %>% 
  dplyr::mutate(sample_id_matching = gsub(Barcode_ID, pattern = c("_3p"), replacement = "")) %>% 
  dplyr::mutate(Cell_category = case_when(Cell_type == "iPSC" ~ "iPSC",
                                          Cell_type == "Neuroepithelial" ~ "Neuroepithelial",
                                          Cell_type == "Neural progenitors" ~ "Neural progenitor",
                                          Cell_type == "Neuron" ~ "Neuron",
                                          Cell_type == "Astrocyte" & Treatment == "None" ~ "Astrocyte (untreated)",
                                          Cell_type == "Astrocyte" & Treatment != "None" ~ "Astrocyte (treated)",
                                          Cell_type == "Microglia" & Treatment == "None" ~ "Microglia (untreated)",
                                          Cell_type == "Microglia" & Treatment != "None" ~ "Microglia (treated)")) %>% 
  data.frame()

# reads per sample generated by cDNA_cupcake
reads_per_sample <- 
  read_csv(args$iPSC$path_to_reads,
           show_col_types = F) 

# Functions ---------------------------------------------------------------

# Select samples to be included and calculate total reads based on remaining samples
Samples_and_total_reads <-
  function(read_data, samples, include) {
    
    
    if(include == TRUE) {
      
      final_reads <-
        read_data %>% 
        dplyr::select(1, contains(samples)) %>% 
        dplyr::mutate(Clontech_5p..Total_reads_3p = rowSums(.[2:ncol(.)]))
      
    } else {
      
      final_reads <-
        read_data %>% 
        dplyr::select(1, !contains(samples)) %>% 
        dplyr::mutate(Clontech_5p..Total_reads_3p = rowSums(.[2:ncol(.)]))
      
    }
    
    return(final_reads)
    
  }

# Filter by gene of interest
Filter_by_gene <-
  function(data, gene) {
    
    Filtered <- data %>% 
      dplyr::filter(associated_gene %in% gene) %>% data.frame()
    
    new_colnames <- colnames(Filtered) %>% 
      gsub(pattern = c("Clontech_5p..|_3p"), replacement = "")
    
    for (i in str_which(colnames(Filtered), pattern = "^Clontech")) {
      
      Filtered[(new_colnames[i])] <- Filtered[i] / colSums(Filtered[i]) *100
    }
    
    return(Filtered)
    
  }

# Get a list with all transcripts overlapping the CDS of interest
# Each element in the list is a query transcript
compare_CDS <-
  function(query, subject, pb_ids) {
    
    # Filter gff files to only contain CDS and split by transcript
    
    query_cds <-
      query[query$type == "CDS" & query$transcript_id %in% pb_ids] %>%
      split(., .$transcript_id)
    
    subject_cds <-
      subject[subject$type == "CDS" & subject$gene_id %in% unique(unlist(query_cds)$gene_id)] %>%
      split(., .$transcript_id)
    
    # create a match list and an index
    
    matches <- list()
    
    for (i in names(query_cds)) {
      
      index <- 0
      
      for (j in 1:length(subject_cds)) {
        
        if (length(query_cds[[i]]) == length(subject_cds[[j]]) &
            all(rle(countOverlaps(query_cds[[i]], subject_cds[[j]], type = "equal"))$values) == 1) {
          
          index <- index + 1
          
          matches[[i]][index] <- names(subject_cds[j])
          
        }
      }
    }
    
    return(
      lapply(matches, data.frame))
    
  }

# Combine list with transcripts per CDS and annotation
combine_TOI_and_annotation <-
  function(TOI, annotation) {
    
    comb_list <-
      lapply(TOI, function(x) {
        left_join(x, annotation, by = c("X..i.." = "isoform")) %>% 
          dplyr::rename("isoform" = "X..i..") 
      }
      )
    
    return(
      bind_rows(comb_list, .id = "CDS"))
  }

# Main --------------------------------------------------------------------

# Select samples to include
reads_per_sample_filtered <-
  Samples_and_total_reads(read_data = reads_per_sample,
                          samples = Samples$Barcode_ID[Samples$Cell_type == "NA"],
                          include = FALSE)

# Add reads to transcript classifications
Isoforms_with_reads <- 
  SQANTI_classification_iPSC %>% 
  dplyr::left_join(., reads_per_sample_filtered, by = c("isoform" = "id")) 

# List with gene annotated elements
annotated_genes <-
  list(Filter_by_gene(data = Isoforms_with_reads, 
                      gene = "ENSG00000177628.16"), 
       Filter_by_gene(data = Isoforms_with_reads, 
                      gene = "ENSG00000160766.14")) %>% bind_rows(!!!.)

# Get a list for each transcript (pb_id) and the matching transcripts from the iPSC data in regards to CDS
Transcripts_of_interest <-
  compare_CDS(query = Isoforms_Brain, 
              subject = Isoforms_iPSC, 
              pb_ids = c("PB.845.2888",
                         # "PB.845.2786",
                         # "PB.845.2627",
                         # "PB.845.2629",
                         "PB.845.2954",
                         "PB.845.525",
                         "PB.845.1693"))

# Final data.frame to plot
Data_to_plot <-
  
  # Combine transcripts of interest with SQANTI annotation, including reads per transcript
  combine_TOI_and_annotation(TOI = Transcripts_of_interest, 
                             annotation = annotated_genes) %>%
  
  # Select columns to use
  dplyr::select(CDS, 73:95)  %>% 
  
  # pivot data
  pivot_longer(!CDS, names_to = "Sample_ID", values_to = "Relative_expression")  %>%
  
  # aggregate data by sum
  aggregate(data = ., Relative_expression ~ CDS + Sample_ID, FUN = sum) %>% 
  
  # add cell type per sample
  left_join(., 
            dplyr::select(Samples, sample_id_matching, Cell_category),
            by = c("Sample_ID" = "sample_id_matching")) %>% 
  
  na.omit() %>% 
  
  # add gene information
  dplyr::mutate(Gene = case_when(CDS %in% c("PB.845.2888",
                                            "PB.845.2954") ~ "GBA",
                                 CDS %in% c("PB.845.525",
                                            "PB.845.1693") ~ "GBAP1")) %>% 
  dplyr::filter(Cell_category %in% c("Neuron", "Astrocyte (untreated)", "Microglia (untreated)")) %>% 
  dplyr::mutate(Cell_category = str_replace(Cell_category, "\\s*\\(untreated\\)", ""))
  

# Plot per gene
for (i in unique(Data_to_plot$Gene)) {
  
  Data_to_plot %>% 
    dplyr::filter(Gene == i) %>% 
    
    ggplot(aes(x = factor(Cell_category, levels = c("Neuron",
                                                    "Astrocyte",
                                                    "Microglia"
                                                    )), 
               y = Relative_expression,
               fill = Cell_category)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    scale_y_continuous(labels = function(x) paste0(x, " %")) +
    scale_fill_manual(breaks = c("Neuron", "Astrocyte", "Microglia"),
                      values = c("#a6cee3", "#7fc564", "#fdbf6f")) +
    labs(x = "",
         y = "Relative expression (%)") +
    facet_wrap(vars(CDS), 
               scales = "free_y", 
               ncol = 3) +
    stat_compare_means(size = 6) +
    theme_classic() +
    theme(legend.position = "none",
          axis.title = element_text(size = 22),
          axis.text.x = element_text(face = "bold",
                                     size = 20, angle = 45, 
                                     hjust = 1),
          axis.text.y = element_text(face = "bold",
                                     size = 16),
          strip.text.x = element_text(face = "bold",
                                      size = 16),
          strip.background =element_rect(fill = "gray80"))
  
  ggsave(filename = paste0(i, "_unique_CDS_detected_iPSC_plot.svg"), 
         path = here::here("results", "transcripts"), 
         width = 14, 
         height = 8, 
         dpi = 600
  )
}
